internal/native_image/rules.bzl

```diff
@@ -146,154 +60,69 @@ def _graal_binary_implementation(ctx):
+    inputs = depset(
+        direct_inputs,
+        transitive = transitive_inputs,
     )
 
-    if ctx.attr.reflection_configuration != None:
-        args.add(ctx.file.reflection_configuration, format = "-H:ReflectionConfigurationFiles=%s")
-        direct_inputs.append(ctx.file.reflection_configuration)
-
-    if ctx.attr.include_resources != None:
-        args.add(ctx.attr.include_resources, format = "-H:IncludeResources=%s")
-
-    if ctx.attr.jni_configuration != None:
-        args.add(ctx.file.jni_configuration, format = "-H:JNIConfigurationFiles=%s")
-        direct_inputs.append(ctx.file.jni_configuration)
-        args.add("-H:+JNI")
-
-    inputs = depset(direct_inputs, transitive = transitive_inputs)
-
     run_params = {
         "outputs": [binary],
-        "arguments": [args],
         "executable": graal,
         "inputs": inputs,
         "mnemonic": "NativeImage",
-        "env": env,
+        "env": native_toolchain.env,
+        "execution_requirements": {k: "" for k in native_toolchain.execution_requirements},
+        "progress_message": "Compiling native image %{label}",
+        "toolchain": Label(_GVM_TOOLCHAIN_TYPE),
     }
 
-    if not ctx.attr._legacy_rule:
-        run_params.update(
-            use_default_shell_env = ctx.attr.enable_default_shell_env,
-            progress_message = "Compiling native image %{label}",
-            toolchain = Label(_GVM_TOOLCHAIN_TYPE),
+    graal_actions = _wrap_actions_for_graal(ctx.actions)
+    if ctx.target_platform_has_constraint(ctx.attr._macos_constraint[platform_common.ConstraintValueInfo]):
+        xcode_args = ctx.actions.args()
+
+        # Bazel passes DEVELOPER_DIR and SDKROOT to every locally executed action that sets the
+        # environment variables passed by apple_support.run. However, Graal sanitizes the
+        # environment and removes these variables if set directly. We need to convert them into
+        # -E options and rely on apple_support's argument replacement to pass them through to the
+        # compiler invoked by Graal.
+        # https://github.com/oracle/graal/blob/77a7f6a691024d22367ae33be4da0c15ceb6a246/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java#L1801-L1808
+        xcode_args.add(apple_support.path_placeholders.xcode(), format = "-EDEVELOPER_DIR=%s")
+        xcode_args.add(apple_support.path_placeholders.sdkroot(), format = "-ESDKROOT=%s")
+        apple_support.run(
+            actions = graal_actions,
+            apple_fragment = ctx.fragments.apple,
+            xcode_config = ctx.attr._xcode_config[apple_common.XcodeVersionConfig],
+            xcode_path_resolve_level = apple_support.xcode_path_resolve_level.args,
+            arguments = [args, xcode_args],
+            **run_params
         )
 
-    ctx.actions.run(
-        **run_params
-    )
+    else:
+        # activate default shell environment on windows and linux
+        run_params["use_default_shell_env"] = True
```

Comment:
This is problematic with Bazel 6 and earlier as it will override the toolchain env: https://github.com/bazelbuild/bazel/issues/5980

The Windows MSVC toolchain configured by Bazel adds `INCLUDE` and `LIB` to the toolchain env, so this shouldn't be needed.

Which failure are you seeing without this?

Reply:
The [failures are here](https://github.com/sgammon/rules_graalvm/actions/runs/6046019848/job/16406932089); Bazel doesn't complain, it's `native-image` itself complaining that MSVC cannot be located:

```
Error: Failed to automatically set up Windows build environment.
Please make sure that Visual Studio 2022 version 17.1.0 or later is installed on your system.
You can download it at https://visualstudio.microsoft.com/downloads/.
If this error persists, please try and run GraalVM Native Image in an x64 Native Tools Command Prompt or file a ticket.
```

[As far as I can tell](https://github.com/oracle/graal/blob/46de6045d403bb373f65d5e3303f9d3d09f838df/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/WindowsBuildEnvironmentUtil.java#L45), `native-image` is looking at the [`PATH` for `cl.exe`](https://github.com/oracle/graal/blob/46de6045d403bb373f65d5e3303f9d3d09f838df/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/WindowsBuildEnvironmentUtil.java#L130-L136) (via `cmd.exe /c where cl.exe`) or otherwise gathering the [via `vcvarsall.bat`](https://github.com/oracle/graal/blob/46de6045d403bb373f65d5e3303f9d3d09f838df/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/WindowsBuildEnvironmentUtil.java#L99-L100) and using some [env vars](https://github.com/oracle/graal/blob/46de6045d403bb373f65d5e3303f9d3d09f838df/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/WindowsBuildEnvironmentUtil.java#L46).

While [Bazel needs access to certain variables for its own MSVC resolution](https://bazel.build/configure/windows#build_cpp), `native-image` itself actually doesn't care if Bazel can resolve VS (as in, no invocation of VS or `cl.exe` occurs through Bazel). Furthermore, `native-image` expects `--native-compiler-path` to point to `cl.exe`, but Bazel feels free to substitute this with a batch file wrapper. As such, before this PR, [`pass_compiler_path` defaulted to `False` on Windows](https://github.com/sgammon/rules_graalvm/blob/main/graalvm/nativeimage/rules.bzl#L66-L69) so that `native-image` could _instead_ load it from [the shimmed `PATH`](https://github.com/sgammon/rules_graalvm/blob/main/internal/native_image/rules.bzl#L212).

I have observed that withholding `INCLUDE` and `LIB` fail the build, and I'm not entirely sure _why_, but `use_default_shell_env` fixes this, [even on Bazel 6](https://github.com/sgammon/rules_graalvm/actions/runs/6047230764/job/16410282426?pr=80) (that is the case that should be failing, right?).

Reply:
Ah, wait, it looks like I was still including these in the default env:

https://github.com/sgammon/rules_graalvm/blob/0f5d42b9f6b50d26b71ded92ea6498cb29c3decc/example/integration_tests/bazel6/.bazelrc#L1-L4

Okay. Let me roll back the `use_default_shell_env` change and drop those variables and see where it leaves us.